### PR TITLE
lib/nolibc: Import `strstr` from `musl`

### PR DESCRIPTION
**Pull request by mogasergiu**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A

### Description of changes

Import from musl the basic implementation of the widely available string searching method `strstr`.

### CODE REVIEW

File: lib/nolibc/musl-imported/src/string/strstr.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
+/* taken from musl */
+
+#include <string.h>
+#include <stdint.h>
+
+static char *twobyte_strstr(const unsigned char *h, const unsigned char *n)
+{
+	uint16_t nw = n[0]<<8 | n[1], hw = h[0]<<8 | h[1];
+	for (h++; *h && hw != nw; hw = hw<<8 | *++h);
```
Comment (RaduNichita) [inline lib/nolibc/musl-imported/src/string/strstr.c]: Hi @mogasergiu ,

Can you please fix the linting problem?

File: lib/nolibc/musl-imported/src/string/strstr.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
+/* taken from musl */
+
+#include <string.h>
+#include <stdint.h>
+
+static char *twobyte_strstr(const unsigned char *h, const unsigned char *n)
```
Comment (RaduNichita) [inline lib/nolibc/musl-imported/src/string/strstr.c:6]: Can we refactor those function to make something more generic?

File: lib/nolibc/musl-imported/src/string/strstr.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
+/* taken from musl */
```
Comment (RaduNichita) [inline lib/nolibc/musl-imported/src/string/strstr.c:1]: Could you provide a link from where did you take this?

File: lib/nolibc/musl-imported/src/string/strstr.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+static char *fourbyte_strstr(const unsigned char *h, const unsigned char *n)
+{
+	uint32_t nw = (uint32_t)n[0]<<24 | n[1]<<16 | n[2]<<8 | n[3];
+	uint32_t hw = (uint32_t)h[0]<<24 | h[1]<<16 | h[2]<<8 | h[3];
+	for (h+=3; *h && hw != nw; hw = hw<<8 | *++h);
+	return *h ? (char *)h-3 : 0;
+}
+
+#define MAX(a,b) ((a)>(b)?(a):(b))
```
Comment (RaduNichita) [inline lib/nolibc/musl-imported/src/string/strstr.c]: Should we define this? I don't think it is a good idea to put this here.

**Review by RaduNichita — COMMENTED** [review]
Thanks @mogasergiu 💯  

Left some questions!

File: lib/nolibc/musl-imported/src/string/strstr.c (hunk 3 above)
Comment (mogasergiu) [reply lib/nolibc/musl-imported/src/string/strstr.c:1]: I simply aligned this with `strsignal.c` also taken from musl.  I don't think it is needed, check our `strsignal.c` file.

File: lib/nolibc/musl-imported/src/string/strstr.c (hunk 4 above)
Comment (mogasergiu) [reply lib/nolibc/musl-imported/src/string/strstr.c]: This is directly taken from musl's [strstr.c](https://github.com/bminor/musl/blob/master/src/string/strstr.c).

File: lib/nolibc/musl-imported/src/string/strstr.c (hunk 2 above)
Comment (mogasergiu) [reply lib/nolibc/musl-imported/src/string/strstr.c:6]: I generally agree with your points here. And although yes, musl is under MIT license, allowing a great amound of freedom, including publishing modified versions, I would still keep it as is :thinking: 

EDIT: I will probably re-write some more parts of this file. It is after all a copy-paste of an MIT licensed file.

File: lib/nolibc/musl-imported/src/string/strstr.c (hunk 1 above)
Comment (mogasergiu) [reply lib/nolibc/musl-imported/src/string/strstr.c]: Yeah this does look like something worth aligning to our coding style standards :+1:.

File: lib/nolibc/musl-imported/src/string/strstr.c (hunk 4 above)
Comment (mogasergiu) [reply lib/nolibc/musl-imported/src/string/strstr.c]: Hmm, now that I take a second look at `strsignal.c` it does seem to be quite modified compared to the original. I will come back to this PR and re-write some more parts to look better.

**Comment by mogasergiu** [comment]
Hey @RaduNichita! Thanks for not letting these slip by :muscle:! Did some checkpatch fixes, but I don't think we should change more than that as I want to be able to easily update this source file w.r.t. to musl upstream if it ever comes to it, so I wouldn't change too much of it functionality wise. But, in general, I agree with you, a lot could be refactored to be better here.

**Review by flpostolache — APPROVED** [review]
Looks good to me!

Reviewed-by: Florin Postolache <florin.postolache.of@gmail.com>

**Review by RaduNichita — APPROVED** [review]
All good.

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**